rpc/client/localclient.go

```diff
@@ -48,106 +56,177 @@ var (
+	return core.Block(&rpctypes.Context{}, height)
 }
 
 func (Local) BlockResults(height *int64) (*ctypes.ResultBlockResults, error) {
-	return core.BlockResults(height)
+	return core.BlockResults(&rpctypes.Context{}, height)
 }
 
 func (Local) Commit(height *int64) (*ctypes.ResultCommit, error) {
-	return core.Commit(height)
+	return core.Commit(&rpctypes.Context{}, height)
 }
 
 func (Local) Validators(height *int64) (*ctypes.ResultValidators, error) {
-	return core.Validators(height)
+	return core.Validators(&rpctypes.Context{}, height)
 }
 
 func (Local) Tx(hash []byte, prove bool) (*ctypes.ResultTx, error) {
-	return core.Tx(hash, prove)
+	return core.Tx(&rpctypes.Context{}, hash, prove)
 }
 
 func (Local) TxSearch(query string, prove bool, page, perPage int) (*ctypes.ResultTxSearch, error) {
-	return core.TxSearch(query, prove, page, perPage)
+	return core.TxSearch(&rpctypes.Context{}, query, prove, page, perPage)
 }
 
-func (c *Local) Subscribe(ctx context.Context, subscriber string, query tmpubsub.Query, outCapacity ...int) (types.Subscription, error) {
-	return c.EventBus.Subscribe(ctx, subscriber, query, outCapacity...)
-}
+// Subscribe implements EventsClient by using local eventBus to subscribe given
+// subscriber to query.By default, returns a channel with cap=1. Error is
+// returned if it fails to subscribe.
+// Channel is never closed to prevent clients from seeing an erroneus event.
+func (c *Local) Subscribe(ctx context.Context, subscriber, query string, outCapacity ...int) (out <-chan ctypes.ResultEvent, err error) {
+	q, err := tmquery.New(query)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to parse query")
+	}
+	sub, err := c.EventBus.Subscribe(ctx, subscriber, q)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to subscribe")
+	}
+
+	outCap := 1
+	if len(outCapacity) > 0 {
+		outCap = outCapacity[0]
+	}
 
-func (c *Local) Unsubscribe(ctx context.Context, subscriber string, query tmpubsub.Query) error {
-	return c.EventBus.Unsubscribe(ctx, subscriber, query)
+	outc := make(chan ctypes.ResultEvent, outCap)
+	go func(sub types.Subscription) {
+		for {
+			select {
+			case msg := <-sub.Out():
+				result := ctypes.ResultEvent{Query: query, Data: msg.Data(), Tags: msg.Tags()}
+				if cap(outc) == 0 {
+					outc <- result
+				} else {
+					select {
+					case outc <- result:
+					default:
+						c.Logger.Error("wanted to publish ResultEvent, but out channel is full", "result", result, "query", result.Query)
+					}
+				}
+			case <-sub.Cancelled():
+				if sub.Err() != tmpubsub.ErrUnsubscribed {
+					// resubscribe
+					c.Logger.Error("subscription was cancelled, resubscribing...", "err", err, "query", query)
+					var err error
+					for {
+						if !c.IsRunning() {
+							return
+						}
+
+						ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+						defer cancel()
+						sub, err = c.EventBus.Subscribe(ctx, subscriber, q)
+						if err == nil {
+							break
+						}
```

Comment:
exponential timeout?

Reply:
Should we consider closing the entire WS connection if an event is missed and pushing this responsibility on the client?